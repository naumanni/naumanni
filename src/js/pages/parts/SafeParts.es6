import React from 'react'

import {
  TOKEN_TEXT, TOKEN_BREAK, TOKEN_URL, TOKEN_MENTION, TOKEN_HASHTAG, TOKEN_EMOJI,
} from 'src/constants'


const SPLIT_URL_REX = /https?:\/\/([^\/]+)/
const MAX_PATH_LENGTH = 15


export function SafeLink({children, href, ...props}) {
  if(!(href.startsWith('http://') || href.startsWith('https://')))
    href = 'javascript:void(0)'

  return (
    <a href={href} {...props}>{children}</a>
  )
}


export function SafeNote({parsedNote}) {
  return React.createElement(
    'p',
    {},
    ...parsedNote.map((token) => {
      const {type} = token

      if(type === TOKEN_TEXT) {
        return <span>{token.text}</span>
      } else if(type === TOKEN_BREAK) {
        return <br />
      } else if(type === TOKEN_URL) {
        return <a href={token.url} rel="nofollow noopener" target="_blank">{token.url}</a>
      } else if(type === TOKEN_MENTION) {
        return <span>@{token.acct}</span>
      } else if(type === TOKEN_HASHTAG) {
        return <span>#{token.tag}</span>
      } else {
        console.error(token)
      }
    })
  )
}


export function SafeContent({parsedContent, onClickHashTag, ...props}) {
  return React.createElement(
    'p',
    {},
    ...parsedContent.map((token) => {
      const {type} = token

      if(type === TOKEN_TEXT) {
        return <span>{token.text}</span>
      } else if(type === TOKEN_BREAK) {
        return <br />
      } else if(type === TOKEN_URL) {
        const url = new URL(token.url)

        let host = url.host
        let path = `${url.pathname}${url.search}${url.hash}`

        if(url.username || url.password)
          host = `${url.username}:${url.password}@${host}`
        if(path.length > MAX_PATH_LENGTH)
          path = `${path.substring(0, MAX_PATH_LENGTH)}â€¦`

        return (
          <a href={url.href} rel="nofollow noopener" target="_blank">{host + path}</a>
        )
      } else if(type === TOKEN_MENTION) {
        const {acct} = token
        return <a href={_urlFromAcct(acct)} onClick={(e) => props.onAvatarClicked(acct, e)}>@{acct}</a>
      } else if(type === TOKEN_HASHTAG) {
        return <a onClick={onClickHashTag.bind(this, token.tag)}>#{token.tag}</a>
      } else if(type === TOKEN_EMOJI) {
        const {
          draggable = false,
          class: className = '',
          alt = '',
          title = '',
          src,
        } = token
        const props = {className, draggable, alt, title, src}
        return <img {...props} />
      } else {
        console.error(token)
      }
    })
  )
}


function _urlFromAcct(acct) {
  const [username, host] = acct.split('@')
  return `https://${host}/@${username}`
}
