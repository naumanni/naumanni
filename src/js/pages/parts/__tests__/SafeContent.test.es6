import React from 'react'
import {shallow} from 'enzyme'
import renderer from 'react-test-renderer'
import {spy} from 'sinon'

import {
  TOKEN_TEXT, TOKEN_BREAK, TOKEN_URL, TOKEN_MENTION, TOKEN_HASHTAG, TOKEN_EMOJI,
} from 'src/constants'
import {SafeContent} from '../SafeParts'


const takeSnapshot = (component) => {
  const tree = renderer.create(component).toJSON()
  expect(tree).toMatchSnapshot()
}

describe('SafeContent', () => {
  let parsedContent
  let component

  describe('TOKEN_TEXT', () => {
    beforeAll(() => {
      parsedContent = [
        {type: TOKEN_TEXT, text: 'hello'},
      ]
      component = <SafeContent parsedContent={parsedContent} />
    })
    it('match snapshot', () => {
      takeSnapshot(component)
    })
  })

  describe('TOKEN_BREAK', () => {
    beforeAll(() => {
      parsedContent = [
        {type: TOKEN_BREAK},
      ]
      component = <SafeContent parsedContent={parsedContent} />
    })
    it('match snapshot', () => {
      takeSnapshot(component)
    })
  })

  describe('TOKEN_URL', () => {
    describe('normal url', () => {
      beforeAll(() => {
        parsedContent = [
          {type: TOKEN_URL, url: 'http://example.com'},
        ]
        component = <SafeContent parsedContent={parsedContent} />
      })
      it('match snapshot', () => {
        takeSnapshot(component)
      })
    })

    describe('username:password specified url', () => {
      beforeAll(() => {
        parsedContent = [
          {type: TOKEN_URL, url: 'http://john:pass@example.com'},
        ]
        component = <SafeContent parsedContent={parsedContent} />
      })
      it('match snapshot', () => {
        takeSnapshot(component)
      })
    })

    describe('query and anchor specified url', () => {
      beforeAll(() => {
        parsedContent = [
          {type: TOKEN_URL, url: 'http://example.com?q=p#anchor'},
        ]
        component = <SafeContent parsedContent={parsedContent} />
      })
      it('match snapshot', () => {
        takeSnapshot(component)
      })
    })

    describe('long path url', () => {
      beforeAll(() => {
        parsedContent = [
          {type: TOKEN_URL, url: 'http://example.com/abcdefghijklmnopqrstuvwxyz'},
        ]
        component = <SafeContent parsedContent={parsedContent} />
      })
      it('match snapshot', () => {
        takeSnapshot(component)
      })
    })
  })

  describe('TOKEN_MENTION', () => {
    let onAvatarClickedSpy

    beforeAll(() => {
      onAvatarClickedSpy = spy()
      parsedContent = [
        {type: TOKEN_MENTION, acct: 'glpt@friends.nico'},
      ]
      component = <SafeContent parsedContent={parsedContent} onAvatarClicked={onAvatarClickedSpy} />
    })
    it('interact with onAvatarClicked handler', () => {
      const wrapper = shallow(component)
      const link = wrapper.find('a')
      link.simulate('click')
      expect(onAvatarClickedSpy.calledOnce).toBeTruthy()

      const acct = parsedContent[0].acct
      expect(onAvatarClickedSpy.calledWith(acct)).toBeTruthy()
    })
    it('match snapshot', () => {
      takeSnapshot(component)
    })
  })

  describe('TOKEN_HASHTAG', () => {
    const tag = 'naumanni'
    let onClickHashTag

    beforeEach(() => {
      parsedContent = [
        {type: TOKEN_HASHTAG, tag},
      ]
      onClickHashTag = jest.fn()
      component = <SafeContent parsedContent={parsedContent} onClickHashTag={onClickHashTag} />
    })
    it('propagate link click to its onClickHashTag prop', () => {
      const wrapper = shallow(component)
      const link = wrapper.find('a')

      link.simulate('click')

      expect(onClickHashTag).toHaveBeenCalled()
      expect(onClickHashTag).toHaveBeenCalledWith(tag)
    })
    it('match snapshot', () => {
      takeSnapshot(component)
    })
  })

  describe('TOKEN_EMOJI', () => {
    beforeAll(() => {
      parsedContent = [{
        type: TOKEN_EMOJI,
        draggable: false,
        alt: 'üôè',
        title: ':pray:',
        src: '/static/images/emoji/1f606.svg',
      }]
      component = <SafeContent parsedContent={parsedContent} />
    })
    it('match snapshot', () => {
      takeSnapshot(component)
    })
  })
})
