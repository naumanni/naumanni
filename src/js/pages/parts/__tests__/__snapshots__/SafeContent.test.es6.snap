// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SafeContent TOKEN_BREAK match snapshot 1`] = `
<p>
  <br />
</p>
`;

exports[`SafeContent TOKEN_HASHTAG match snapshot 1`] = `
<p>
  <a
    onClick={[Function]}
  >
    #
    naumanni
  </a>
</p>
`;

exports[`SafeContent TOKEN_MENTION match snapshot 1`] = `
<p>
  <a
    href="https://friends.nico/@glpt"
    onClick={[Function]}
  >
    @
    glpt@friends.nico
  </a>
</p>
`;

exports[`SafeContent TOKEN_TEXT match snapshot 1`] = `
<p>
  <span>
    hello
  </span>
</p>
`;

exports[`SafeContent TOKEN_URL long path url match snapshot 1`] = `
<p>
  <a
    href="http://example.com/abcdefghijklmnopqrstuvwxyz"
    rel="nofollow noopener"
    target="_blank"
  >
    example.com/abcdefghijklmnâ€¦
  </a>
</p>
`;

exports[`SafeContent TOKEN_URL normal url match snapshot 1`] = `
<p>
  <a
    href="http://example.com/"
    rel="nofollow noopener"
    target="_blank"
  >
    example.com/
  </a>
</p>
`;

exports[`SafeContent TOKEN_URL query and anchor specified url match snapshot 1`] = `
<p>
  <a
    href="http://example.com/?q=p#anchor"
    rel="nofollow noopener"
    target="_blank"
  >
    example.com/?q=p#anchor
  </a>
</p>
`;

exports[`SafeContent TOKEN_URL username:password specified url match snapshot 1`] = `
<p>
  <a
    href="http://john:pass@example.com/"
    rel="nofollow noopener"
    target="_blank"
  >
    john:pass@example.com/
  </a>
</p>
`;
